engine/src/state_chain_observer/sc_observer/mod.rs

```diff
@@ -619,3 +612,163 @@ where
+	block_hash: state_chain_runtime::Hash,
+	state_chain_client: Arc<StateChainClient>,
+) -> Result<CeremonyIdCounters, anyhow::Error>
+where
+	StateChainClient: SignedExtrinsicApi
+		+ UnsignedExtrinsicApi
+		+ 'static
+		+ Send
+		+ Sync
+		+ super::client::storage_api::StorageApi,
+{
+	// Get the vales of the ceremony id counters at the end of the block
+	let mut ceremony_id_counters = CeremonyIdCounters {
+		ethereum: state_chain_client
+			.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+				state_chain_runtime::Runtime,
+				state_chain_runtime::EthereumInstance,
+			>>(block_hash)
+			.await
+			.context("Failed to get Ethereum CeremonyIdCounter from SC")?,
+
+		polkadot: state_chain_client
+			.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+				state_chain_runtime::Runtime,
+				state_chain_runtime::PolkadotInstance,
+			>>(block_hash)
+			.await
+			.context("Failed to get Polkadot CeremonyIdCounter from SC")?,
+
+		bitcoin: state_chain_client
+			.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+				state_chain_runtime::Runtime,
+				state_chain_runtime::BitcoinInstance,
+			>>(block_hash)
+			.await
+			.context("Failed to get Bitcoin CeremonyIdCounter from SC")?,
+	};
+
+	// Reverse the counters by looking at the events in the block that would of incremented them
+	match state_chain_client
+		.storage_value::<frame_system::Events<state_chain_runtime::Runtime>>(block_hash)
+		.await
+	{
+		Ok(events) => {
+			for event_record in events {
+				match event_record.event {
+					state_chain_runtime::RuntimeEvent::EthereumVault(
+						pallet_cf_vaults::Event::KeygenRequest { .. },
+					) |
+					state_chain_runtime::RuntimeEvent::EthereumThresholdSigner(
+						pallet_cf_threshold_signature::Event::ThresholdSignatureRequest { .. },
+					) => {
+						ceremony_id_counters.ethereum -= 1;
+					},
+
+					state_chain_runtime::RuntimeEvent::PolkadotVault(
+						pallet_cf_vaults::Event::KeygenRequest { .. },
+					) |
+					state_chain_runtime::RuntimeEvent::PolkadotThresholdSigner(
+						pallet_cf_threshold_signature::Event::ThresholdSignatureRequest { .. },
+					) => {
+						ceremony_id_counters.polkadot -= 1;
+					},
+
+					state_chain_runtime::RuntimeEvent::BitcoinVault(
+						pallet_cf_vaults::Event::KeygenRequest { .. },
+					) |
+					state_chain_runtime::RuntimeEvent::BitcoinThresholdSigner(
+						pallet_cf_threshold_signature::Event::ThresholdSignatureRequest { .. },
+					) |
+					state_chain_runtime::RuntimeEvent::BitcoinVault(
+						pallet_cf_vaults::Event::KeyHandoverRequest { .. },
+					) => {
+						ceremony_id_counters.bitcoin -= 1;
+					},
+
+					_ => {
+						// We only care about events that change the Ceremony Id Counter
+					},
+				}
+			}
+		},
+		Err(error) => {
+			bail!("Failed to decode events at block hash {block_hash}. {error}");
+		},
+	}
+
+	Ok(ceremony_id_counters)
+}
+
+pub async fn monitor_p2p_registration_events<StateChainClient, BlockStream: StateChainStreamApi>(
```

Comment:
Shouldn't this be with the p2p code?

Reply:
I couldn't decide. Thought it was nice to keep all the event matching (observing) in the SCO. But if you think it's better in p2p i will move it.